cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,324 @@
+from collections import defaultdict, OrderedDict
+from typing import Optional
+
+from PyQt5.Qt import QTimer, QObject, pyqtSignal
+
+from UM.Logger import Logger
+from UM.Settings import ContainerRegistry
+
+from cura.Machines.ContainerNode import ContainerNode
+
+
+class MaterialGroup:
+    __slots__ = ("name", "root_material_node", "derived_material_node_list")
+
+    def __init__(self, name: str):
+        self.name = name
+        self.root_material_node = None
```

Comment:
In the interest of defensive coding, I suggest that you make the root material node required.

Reply:
This is a bit hard to do. In `MaterialManager`, we try to reconstruct material groups out of the metadata list. Because the container metadata is not ordered in a certain way in the metadata list, we can only try to construct the group along the way and the root material node will be created when it's found. Perhaps we can add a sanity check in MaterialManager to make sure that each MaterialGroup must have a root_material_node?

Reply:
After discussing with you, I went ahead and made it required anyway. Corrected in this commit: 1512a8096b1e7d2e3c570675e42ae7086da32fdd